lib/extras/gain_map.cc

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#include <jxl/gain_map.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+
+#include "lib/jxl/base/byte_order.h"
+#include "lib/jxl/color_encoding_internal.h"
+#include "lib/jxl/dec_bit_reader.h"
+#include "lib/jxl/enc_bit_writer.h"
+#include "lib/jxl/fields.h"
+#include "lib/jxl/memory_manager_internal.h"
+
+namespace {
+
+template <size_t N>
+void* FixedSizeMemoryManagerAlloc(void* opaque, size_t capacity);
+void FixedSizeMemoryManagerFree(void* opaque, void* pointer) {}
+
+template <size_t N>
+class FixedSizeMemoryManager {
```

Comment:
This will only handle one allocation, right? Perhaps we should remember that and return nullptr if trying to allocate again?

Reply:
@sboukortt

Reply:
It’s meant to handle “reallocations”, for when `PaddedBytes` wants to increase its size. This is made possible by the fact that freeing is a no-op, and by the change from `memcpy` to `memmove` in `padded_bytes.h` (since the source and destination will be identical).

Reply:
I guess that's fine, I am not sure I like it too much but I don't see better alternatives...

Reply:
It feels a bit weird that this is defined in `gain_map.cc` and that this kind of functionality is only needed here.

Reply:
Let’s have it here for now and factor it out if we eventually need it elsewhere?